test/runner.js

```diff
@@ -12,46 +12,60 @@ const debug = require( 'debug' )( 'test-runner' ),
 	Mocha = require( 'mocha' ),
 	path = require( 'path' ),
 	program = require( 'commander' ),
+	watchFile = require( 'fs' ).watchFile,
 	chalk = require( 'chalk' );
 
-/**
- * Internal dependencies
- */
-const boot = require( './boot-test' ),
-	setup = require( './setup' );
-
 program
 	.usage( '[options] [files]' )
 	.option( '-R, --reporter <name>', 'specify the reporter to use', 'spec' )
 	.option( '-t, --node-total <n>', 'specify the node total to use', parseInt )
 	.option( '-i, --node-index <n>', 'specify the node index to use', parseInt )
-	.option( '-g, --grep <pattern>', 'only run tests matching <pattern>' );
+	.option( '-g, --grep <pattern>', 'only run tests matching <pattern>' )
+	.option( '-w, --watch', 'watch files for changes' );
 
 program.name = 'runner';
 
 program.parse( process.argv );
 
-const mocha = new Mocha( {
-	ui: 'bdd',
-	reporter: program.reporter
-} );
+const getMocha = function() {
+	/* These internal dependencies are not at the top of the file
+	 * because we need to re-require them every time we invalidate the cache.
+	 *
+	 * We invalidate the cache because Mocha has trouble running a test-suite
+	 * more than once:  https://github.com/mochajs/mocha/issues/995
+	 */
+	Object.keys( require.cache ).forEach( ( key ) => {
+		delete require.cache[ key ];
+	} );
+	const boot = require( './boot-test' );
+	const setup = require( './setup' );
+
+	const mocha = new Mocha( {
+		ui: 'bdd',
+		reporer: program.reporter
+	} );
 
-if ( program.grep ) {
-	mocha.grep( new RegExp( program.grep ) );
-}
+	mocha.suite.beforeAll( boot.before );
+	mocha.suite.afterAll( boot.after );
 
-if ( process.env.CIRCLECI ) {
-	debug( 'Hello Circle!' );
-	// give circle more time by default because containers are slow
-	// why 10 seconds? a guess.
-	mocha.suite.timeout( 10000 );
-}
+	if ( program.grep ) {
+		mocha.grep( new RegExp( program.grep ) );
+	}
+
+	if ( process.env.CIRCLECI ) {
+		debug( 'Hello Circle!' );
+		// give circle more time by default because containers are slow
+		// why 10 seconds? a guess.
+		mocha.suite.timeout( 10000 );
+	}
 
-mocha.suite.beforeAll( boot.before );
-mocha.suite.afterAll( boot.after );
+	testFiles.forEach( setup.addFile );
+	mocha.addFile( path.join( __dirname, 'load-suite.js' ) );
+	return mocha;
+};
 
-files = program.args.length ? program.args : [ process.env.TEST_ROOT ];
-files = files.reduce( ( memo, filePath ) => {
+testFiles = program.args.length ? program.args : [ process.env.TEST_ROOT ];
```

Comment:
We rather should move all logic related to `testFiles` inside `getMocha` to make sure Mocha is able to pick up new files and skip processing removed files.